{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7129e3a6-1b3b-4e85-906e-ed6aaf671d8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO, Seq, Entrez\n",
    "import numpy as np\n",
    "import datetime\n",
    "import ete3\n",
    "import re\n",
    "import math\n",
    "import statistics\n",
    "import random\n",
    "import scipy\n",
    "import h5py\n",
    "import time\n",
    "import os\n",
    "import statsmodels.api as sm\n",
    "\n",
    "import plotly.express as px\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a6a4f8ce-1f7e-469b-b67b-d427080b9803",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_lognorm_CI(theseries, zscore):\n",
    "    theseries = [math.log(x) for x in list(theseries)]\n",
    "    CIlo = statistics.mean(theseries)  + (statistics.stdev(theseries)/2) + \\\n",
    "    zscore * math.sqrt( (statistics.stdev(theseries)/len(theseries)) + ( pow(statistics.stdev(theseries), 2) / (2 * (len(theseries)-1)) ) )\n",
    "\n",
    "    return pow(math.e, CIlo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5c426642-83ba-4cbf-959d-50575e4b5929",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def sum_entropy(entrdf_f, nodedetsdf_f, tfile, outnam):\n",
    "\n",
    "    entrdf = pd.read_csv(entrdf_f,index_col=0)\n",
    "    nodedetsdf = pd.read_csv(nodedetsdf_f)\n",
    "    \n",
    "    #load tree with node labels\n",
    "    thet = ete3.Tree(tfile, format=1)   \n",
    "\n",
    "    print(len(thet))\n",
    "    entrdiffs = []\n",
    "\n",
    "    #fill na for easier processing\n",
    "    nodedetsdf = nodedetsdf.fillna('')\n",
    "    \n",
    "    \n",
    "    counter = 0\n",
    "    #for each node name\n",
    "        #this will be the child node that acquired the mutations\n",
    "    for cnod in thet.traverse():\n",
    "        counter = counter + 1\n",
    "        if counter%100 == 0:\n",
    "            print(counter)#\n",
    "        \n",
    "        \n",
    "        #get node above, this will be the context where the mutations happened\n",
    "        upnod = cnod.up\n",
    "        \n",
    "        if upnod != None:\n",
    "            \n",
    "            if len(upnod.name) > 0:\n",
    "            \n",
    "                #ensure that the node is in the summarised dataframe\n",
    "                if cnod.name in list(nodedetsdf.node):            \n",
    "                    #get all the sites that have been substituted on that branch\n",
    "                    thesites = [x[1:-1] for x in list(nodedetsdf[nodedetsdf.node == cnod.name].subs)[0].split(',') if len(x)>=3]\n",
    "\n",
    "                    for site in thesites:\n",
    "\n",
    "                        #list of entropies for that site excluding the parent name\n",
    "                        nodexcl_entr = entrdf.loc[[x for x in list(entrdf.index) if x != upnod.name], site]\n",
    "                        nodexcl_entr.dropna(inplace=True)\n",
    "\n",
    "                        #need at least 3 seqs that don't have gaps (informative entr)\n",
    "                        if len(nodexcl_entr) > 2:\n",
    "\n",
    "                            entrmed = nodexcl_entr.median()\n",
    "                            entrCI = calc_lognorm_CI(nodexcl_entr, 2.33)\n",
    "                            # print(n.name, site)\n",
    "                            siteentr = entrdf.loc[upnod.name, site]\n",
    "                            entrdiffs.append([upnod.name, cnod.name, site, siteentr, entrmed, entrCI])\n",
    "    \n",
    "    rezdf = pd.DataFrame(entrdiffs)\n",
    "    rezdf.columns = ['node', 'child', 'site', 'site-entr', 'median-entr', 'CI99-entr']\n",
    "\n",
    "    #drop NA on entropy column -> these are indel regions where there's gaps and miscalled indel substitutions\n",
    "    rezdf.dropna(subset='site-entr', inplace = True)\n",
    "    \n",
    "    rezdf.sort_values(by='node', inplace=True)\n",
    "\n",
    "    \n",
    "    #add entropies of random sites excluding the true ones\n",
    "    randentr_l = []\n",
    "    randmedentr_l = []\n",
    "\n",
    "    for n in sorted(list(set(rezdf.node))):\n",
    "        \n",
    "        truesites = list(rezdf[rezdf.node == n].site)\n",
    "        #dropna here -> i.e. exclude sites with indels in the whole alignment\n",
    "        thisnentrdf = entrdf.dropna(axis=1)\n",
    "        thisnentrdf = thisnentrdf[thisnentrdf.index == n]\n",
    "\n",
    "        #randomly get an equal number of random sites that did not get substituted\n",
    "        randsites =  random.sample([x for x in list(thisnentrdf.columns) if x not in truesites], len(truesites))\n",
    "    \n",
    "        for s in randsites:\n",
    "            \n",
    "            #get the entropy of the random site selected\n",
    "            randentr_l.append(entrdf.loc[n, s])\n",
    "\n",
    "            #list of entropies for that site excluding the parent name\n",
    "            nodexcl_randentr = entrdf.loc[[x for x in list(entrdf.index) if x != n], s]\n",
    "            nodexcl_randentr.dropna(inplace=True)\n",
    "            #add the median for that site\n",
    "            randmedentr_l.append(nodexcl_randentr.median())\n",
    "\n",
    "    rezdf['rand-entr'] = randentr_l\n",
    "    rezdf['rand-entr-med'] = randmedentr_l\n",
    "\n",
    "    \n",
    "    rezdf['cat'] = ['internal' if 'NODE_' in x else 'external' for x in list(rezdf.child)]\n",
    "    rezdf['med-diff'] = rezdf['site-entr'] -  rezdf['median-entr']\n",
    "    rezdf['CI99-diff'] = rezdf['site-entr'] -  rezdf['CI99-entr']\n",
    "    \n",
    "    rezdf['rand-med-diff'] = rezdf['rand-entr'] - rezdf['rand-entr-med']\n",
    "\n",
    "    rezdf.to_csv(outnam, index=False)\n",
    "    \n",
    "    return rezdf\n",
    "                    \n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6883b8de",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_meddiffcomp(entrrezdf):\n",
    "    entr_HAall_rVsv = entrrezdf.melt(value_vars=['med-diff', 'rand-med-diff'], value_name='entropy difference from the median')\n",
    "    entr_HAall_rVsv = entr_HAall_rVsv.replace({'med-diff':'substituted site', 'rand-med-diff':'random site'})\n",
    "    \n",
    "    therange = [min(entrrezdf['med-diff'].quantile(0.25) , entrrezdf['rand-med-diff'].quantile(0.25))-0.1,\n",
    "                max(entrrezdf['med-diff'].quantile(0.75) , entrrezdf['rand-med-diff'].quantile(0.75))+0.1]\n",
    "    print(therange)\n",
    "    \n",
    "    fig = px.box(entr_HAall_rVsv, y='entropy difference from the median', color='variable', points=False, range_y=therange, color_discrete_sequence = palette5 )\n",
    "#     fig = px.box(entr_HAall_rVsv, y='entropy difference from the median', color='variable', points=False, color_discrete_sequence = palette5 )\n",
    "    fig.add_hline(0, line_width=1, line_dash=\"dash\", line_color=\"grey\")\n",
    "    fig.update_layout({'plot_bgcolor': 'rgba(0, 0, 0, 0)', 'paper_bgcolor': 'rgba(0, 0, 0, 0)'})\n",
    "    fig.update_traces(boxmean=True)\n",
    "     \n",
    "    fig.show()   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29060fe3",
   "metadata": {},
   "source": [
    "<br>\n",
    "<br>\n",
    "\n",
    "## Summarise pLM entropies per substituted site\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e10ba37b-d3d4-4c98-9b78-919c894ea015",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "for sero in ['gisaid_h7_270624_filt', 'gisaid_h5_270624_filt', 'ncbi_h1_110424_filt', 'ncbi_h3n2_110424_filt']:\n",
    "\n",
    "    print(sero)\n",
    "    \n",
    "    for entrfile in os.listdir('data/%s/site-entropy-aln/'%sero):\n",
    "        \n",
    "        mod = entrfile.replace('%s-'%sero, '').replace('-site_entropy_aln.csv', '')\n",
    "    \n",
    "        print('\\n' + mod)\n",
    "        \n",
    "        sum_entropy('data/%s/site-entropy-aln/%s'%(sero,entrfile), \n",
    "                'data/%s/%s_nodedetails.csv'%(sero,sero),\n",
    "                'data/%s/%s_tf.nwk'%(sero,sero),\n",
    "                'data/%s/entr-rez/%s_tf-%s_entr-rez.csv'%(sero, sero, mod))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2c749fd",
   "metadata": {},
   "source": [
    "<br>\n",
    "<br>\n",
    "\n",
    "## Plotting difference between substituted and random sites' pLM entropies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6013ce9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_meddiffcomp(entrrezdf, outf):\n",
    "    entr_HAall_rVsv = entrrezdf.melt(value_vars=['med-diff', 'rand-med-diff'], value_name='entropy difference from the median')\n",
    "    entr_HAall_rVsv = entr_HAall_rVsv.replace({'med-diff':'substituted site', 'rand-med-diff':'random site'})\n",
    "    \n",
    "    therange = [min(entrrezdf['med-diff'].quantile(0.25) , entrrezdf['rand-med-diff'].quantile(0.25))-0.1,\n",
    "                max(entrrezdf['med-diff'].quantile(0.75) , entrrezdf['rand-med-diff'].quantile(0.75))+0.1]\n",
    "\n",
    "    qdiff = 0\n",
    "    #if sub is fully contained in the nonsub qdiff = 100\n",
    "    if (entrrezdf['rand-med-diff'].quantile(0.75) > entrrezdf['med-diff'].quantile(0.75)) and (entrrezdf['rand-med-diff'].quantile(0.25) < entrrezdf['med-diff'].quantile(0.25)):\n",
    "        qdiff = 100\n",
    "    elif entrrezdf['med-diff'].quantile(0.25) > entrrezdf['rand-med-diff'].quantile(0.75):\n",
    "        qdiff = 0\n",
    "    else:\n",
    "        #the percentage of the med-diff IQR that's under the rand-med-diff 75% IQR point\n",
    "        qdiff = (entrrezdf['rand-med-diff'].quantile(0.75) - entrrezdf['med-diff'].quantile(0.25)) / (entrrezdf['med-diff'].quantile(0.75) - entrrezdf['med-diff'].quantile(0.25))\n",
    "        qdiff = int(qdiff*100)\n",
    "    \n",
    "    print('Difference between top and bottom quartile: ',\n",
    "          qdiff,\n",
    "          entrrezdf['med-diff'].mean() - entrrezdf['rand-med-diff'].mean(), \n",
    "         scipy.stats.skew(entrrezdf['med-diff']),\n",
    "         scipy.stats.skew(entrrezdf['rand-med-diff']))\n",
    "    \n",
    "    \n",
    "    fig = px.box(entr_HAall_rVsv, y='entropy difference from the median', color='variable', points=False, range_y=therange, color_discrete_sequence = palette5 )\n",
    "    fig.add_hline(0, line_width=1, line_dash=\"dash\", line_color=\"grey\")\n",
    "    fig.update_layout(\n",
    "        {'plot_bgcolor': 'rgba(0, 0, 0, 0)', 'paper_bgcolor': 'rgba(0, 0, 0, 0)'},\n",
    "        showlegend=False,\n",
    "        title={ 'text': str(qdiff) + \" %\", 'x': 0.5, 'xanchor': 'center' },\n",
    "        yaxis_title = None)\n",
    "    \n",
    "    fig.update_traces(boxmean=True)\n",
    "    fig.write_image(outf, width=300, height=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efdd304d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for ser in ['gisaid_h7_270624_filt', 'gisaid_h5_270624_filt', 'ncbi_h1_110424_filt', 'ncbi_h3n2_110424_filt']:\n",
    "    \n",
    "    for x in os.listdir('data/%s/entr-rez/'%ser):\n",
    "\n",
    "        if ('H1' not in x) and ('H5' not in x) and ('H7' not in x) and ('H3' not in x) and ('8020' not in x):\n",
    "            print(x)\n",
    "            entrrezdf = pd.read_csv('data/%s/entr-rez/'%ser + x)\n",
    "            plot_meddiffcomp(entrrezdf, 'plotting/fig3/' + x.replace('_entr-rez.csv', '_boxes.svg'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3df873ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_meddiffcomp80(entrrezdf, outf, datfile):\n",
    "    \n",
    "    metadf = pd.read_csv(datfile)\n",
    "    entrrezdf80 = entrrezdf[entrrezdf.node.isin(metadf[metadf.date>2015.16].node)]\n",
    "    \n",
    "    entr_HAall_rVsv = entrrezdf80.melt(value_vars=['med-diff', 'rand-med-diff'], value_name='entropy difference from the median')\n",
    "    entr_HAall_rVsv = entr_HAall_rVsv.replace({'med-diff':'substituted site', 'rand-med-diff':'random site'})\n",
    "    \n",
    "    therange = [min(entrrezdf80['med-diff'].quantile(0.25) , entrrezdf80['rand-med-diff'].quantile(0.25))-0.1,\n",
    "                max(entrrezdf80['med-diff'].quantile(0.75) , entrrezdf80['rand-med-diff'].quantile(0.75))+0.1]\n",
    "   \n",
    "    qdiff = 0\n",
    "    #if sub is fully contained in the nonsub qdiff = 100\n",
    "    if (entrrezdf80['rand-med-diff'].quantile(0.75) > entrrezdf80['med-diff'].quantile(0.75)) and (entrrezdf80['rand-med-diff'].quantile(0.25) < entrrezdf80['med-diff'].quantile(0.25)):\n",
    "        qdiff = 100\n",
    "    elif entrrezdf80['med-diff'].quantile(0.25) > entrrezdf80['rand-med-diff'].quantile(0.75):\n",
    "        qdiff = 0\n",
    "    else:\n",
    "        #the percentage of the med-diff IQR that's under the rand-med-diff 75% IQR point\n",
    "        qdiff = (entrrezdf80['rand-med-diff'].quantile(0.75) - entrrezdf80['med-diff'].quantile(0.25)) / (entrrezdf80['med-diff'].quantile(0.75) - entrrezdf80['med-diff'].quantile(0.25))\n",
    "        qdiff = int(qdiff*100)\n",
    "    \n",
    "    \n",
    "    print('Difference between top and bottom quartile: ', qdiff)\n",
    "    \n",
    "    \n",
    "    fig = px.box(entr_HAall_rVsv, y='entropy difference from the median', color='variable', points=False, range_y=therange, color_discrete_sequence = palette5 )\n",
    "    fig.add_hline(0, line_width=1, line_dash=\"dash\", line_color=\"grey\")\n",
    "    fig.update_layout(\n",
    "        {'plot_bgcolor': 'rgba(0, 0, 0, 0)', 'paper_bgcolor': 'rgba(0, 0, 0, 0)'},\n",
    "        showlegend=False,\n",
    "        title={ 'text': str(qdiff) + \" %\", 'x': 0.5, 'xanchor': 'center' },\n",
    "        yaxis_title = None)\n",
    "    \n",
    "    fig.update_traces(boxmean=True)\n",
    "    fig.write_image(outf, width=300, height=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bc28291",
   "metadata": {},
   "outputs": [],
   "source": [
    "for ser in ['gisaid_h7_270624_filt', 'gisaid_h5_270624_filt', 'ncbi_h1_110424_filt', 'ncbi_h3n2_110424_filt']:\n",
    "    \n",
    "    for x in os.listdir('data/%s/entr-rez/'%ser):\n",
    "\n",
    "        if '8020' in x:\n",
    "            print(x)\n",
    "            entrrezdf = pd.read_csv('data/%s/entr-rez/'%ser + x)\n",
    "            plot_meddiffcomp80(entrrezdf, 'plotting/fig3/' + x.replace('_entr-rez.csv', '_boxes.svg'), 'data/%s/%s_nodedetails.csv'%(ser, ser))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "852e7489",
   "metadata": {},
   "source": [
    "<br>\n",
    "<br>\n",
    "\n",
    "## Logistic regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "5756edfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#do original model and HA-all\n",
    "\n",
    "\n",
    "for sero in ['gisaid_h7_270624_filt',\n",
    "             'gisaid_h5_270624_filt',\n",
    "             'ncbi_h1_110424_filt',\n",
    "             'ncbi_h3n2_110424_filt']:\n",
    "\n",
    "    for mod in [\n",
    "        'T5_uniref_HA_221124',\n",
    "        'T5_UniRef_e2',\n",
    "        'esm2_t33_e10',\n",
    "        'esm2_t33']:\n",
    "\n",
    "        #read the file with all pLM entropies\n",
    "        entrdf = pd.read_csv('data/%s/site-entropy-aln/%s-ASR-%s-site_entropy_aln.csv'%(sero, sero, mod), index_col=0)\n",
    "        #and the file with the substituted nodes\n",
    "        entrezdf = pd.read_csv('data/%s/entr-rez/%s_tf-ASR-%s_entr-rez.csv'%(sero, sero, mod))\n",
    "\n",
    "        logregrez = []\n",
    "\n",
    "        #for each sequence:\n",
    "        for x in set(entrezdf.node):\n",
    "\n",
    "            #get the sites substituted in that internal node sequence\n",
    "            subsites = set(entrezdf[entrezdf.node ==x].site)\n",
    "\n",
    "            #only run the regression if there are more than 3 sites substituted\n",
    "            if len(subsites)>2:\n",
    "\n",
    "                #get pLM entropies for all sites of that sequence\n",
    "                thisseqdf = entrdf.loc[x]\n",
    "                thisseqdf = pd.DataFrame(thisseqdf.T)\n",
    "\n",
    "                #annotate substituted (1) and non-substituted sites (0)\n",
    "                thisseqdf['sub'] = [1 if int(n) in subsites else 0 for n in list(thisseqdf.index)]\n",
    "\n",
    "                thisseqdf.columns = ['entr', 'sub']\n",
    "\n",
    "                thisseqdf.dropna(subset='entr', inplace=True)\n",
    "\n",
    "                #fit logistic regression\n",
    "                X = thisseqdf['entr']\n",
    "                y = thisseqdf['sub']\n",
    "                X = sm.add_constant(X)\n",
    "\n",
    "                model = sm.Logit(y,X)\n",
    "                result = model.fit(method='bfgs')\n",
    "\n",
    "                pseudo_r2 = 1 - (result.llf / result.llnull)    \n",
    "\n",
    "                slope = result.params['entr']\n",
    "\n",
    "                logregrez.append([x, result.llr_pvalue, pseudo_r2, slope])\n",
    "\n",
    "        logregdf = pd.DataFrame(logregrez)\n",
    "\n",
    "        logregdf.columns = ['node', 'pval', 'pseudoR2', 'slope']\n",
    "\n",
    "        logregdf.to_csv('data/%s/logreg/%s_tf-ASR-%s_logreg-pernode-rez.csv'%(sero, sero, mod),index=False)\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "47efe300",
   "metadata": {},
   "outputs": [],
   "source": [
    "#do alignment\n",
    "\n",
    "\n",
    "for sero in ['gisaid_h7_270624_filt',\n",
    "             'gisaid_h5_270624_filt',\n",
    "             'ncbi_h1_110424_filt',\n",
    "             'ncbi_h3n2_110424_filt']:\n",
    "\n",
    "    seqdic = SeqIO.to_dict(SeqIO.parse('data/%s/ancestral_sequences.fasta'%sero, 'fasta'))\n",
    "    \n",
    "    #remove internal node seqs from per site entropy calculations\n",
    "    aldf = pd.DataFrame({x:[a for a in seqdic[x]] for x in list(seqdic) if 'NODE_' not in x})\n",
    "    \n",
    "    entrezdf = pd.read_csv('data/%s/entr-rez/%s_tf-ASR-esm2_t33_e10_entr-rez.csv'%(sero, sero))\n",
    "    \n",
    "    \n",
    "    #make MSA entropy list\n",
    "    alentropy = []\n",
    "    \n",
    "    for i in range(len(aldf)):\n",
    "        reslist = aldf.iloc[i].tolist()\n",
    "    #     print(reslist)\n",
    "        if '-' in reslist:\n",
    "            reslist.remove('-')\n",
    "        proplist = [reslist.count(x)/len(reslist) for x in set(reslist)]\n",
    "        #calculate alignment site entropy\n",
    "        alentropy.append(scipy.stats.entropy(proplist, base=2))\n",
    "\n",
    "        \n",
    "        \n",
    "\n",
    "    logregrez = []\n",
    "\n",
    "    #for each sequence:\n",
    "    for x in set(entrezdf.node):\n",
    "\n",
    "        subsites = set(entrezdf[entrezdf.node ==x].site)\n",
    "\n",
    "        if len(subsites)>2:\n",
    "\n",
    "            thisseqdf = pd.DataFrame(alentropy)\n",
    "\n",
    "            thisseqdf['sub'] = [1 if int(n) in subsites else 0 for n in list(thisseqdf.index)]\n",
    "\n",
    "            thisseqdf.columns = ['entr', 'sub']\n",
    "\n",
    "            thisseqdf.dropna(subset='entr', inplace=True)\n",
    "\n",
    "    #         print(subsites)\n",
    "    #         print(thisseqdf.sort_values(by='sub'))\n",
    "\n",
    "\n",
    "            #regression\n",
    "            X = thisseqdf['entr']\n",
    "            y = thisseqdf['sub']\n",
    "            X = sm.add_constant(X)\n",
    "\n",
    "            model = sm.Logit(y,X)\n",
    "            result = model.fit(method='bfgs')\n",
    "\n",
    "            pseudo_r2 = 1 - (result.llf / result.llnull)    \n",
    "\n",
    "    #         thisseqdf['logreg'] = result.predict(X)\n",
    "\n",
    "    #         fig = px.scatter(thisseqdf, x='entr', y=['sub', 'logreg'] )\n",
    "    #         fig.show()\n",
    "\n",
    "            slope = result.params['entr']\n",
    "\n",
    "            logregrez.append([x, result.llr_pvalue, pseudo_r2, slope])\n",
    "\n",
    "    logregdf = pd.DataFrame(logregrez)\n",
    "\n",
    "    logregdf.columns = ['node', 'pval', 'pseudoR2', 'slope']\n",
    "\n",
    "    logregdf.to_csv('data/%s/logreg/%s_tf-MSA_logreg-pernode-rez.csv'%(sero, sero),index=False)\n",
    "\n",
    "    \n",
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
